### PR TITLE
[#103325898] Strip whitespace from submitted form values

### PR DESCRIPTION
**Pull request by pcraig3**
This pull request adds a whitespace-stripping WTF Field that we can use instead of a string field, and it strips surrounding whitespace from input submitted through our content-built forms.

[>> Story in Pivotal](https://www.pivotaltracker.com/story/show/103325898)

***

### Content Builder Data

[`ImmutableMultiDict`](http://werkzeug.pocoo.org/docs/0.10/datastructures/#werkzeug.datastructures.ImmutableMultiDict)s can be converted into a regular dict (with duplicate keys), as well as created from them.
So I've written a leading/trailing whitespace-stripping function to recursively `strip()` all values of a `dict`, ( or `list` or `tuple`, or `str`).  Not sure if this is the best solution, but it looks cool.

It addresses the problem of suppliers in their declaration copy+pasting emails with surrounding whitespace that then fail to validate with an unhelpful error message, [as pointed out](https://github.com/alphagov/digitalmarketplace-supplier-frontend/pull/297#issuecomment-144679574) by @quis.

### WTF Field

<sub>Most of the explanation I've just cribbed from my other open pull request [in the supplier app](https://github.com/alphagov/digitalmarketplace-supplier-frontend/pull/297#issuecomment-144679574).</sub>

I've subclassed the WTF `StringField` class with one that preprocesses all user-inputted values.  Specifically, all submitted values will be stripped of leading and trailing whitespace.

What happens if I'm trying to login with my valid email and password:

Submitted values | Processed values | Result
------------ | ------------ | -------------
`'email@email.com'`, `'password'` | `'email@email.com'`, `'password'` | :white_check_mark:
`' email@email.com '`, `'password'` | `'email@email.com'`, `'password'` | :white_check_mark:
`'email@email.com'`, `' password '` | `'email@email.com'`, `' password '` | :x:
`' email@email.com '`, `' password '` | `'email@email.com'`, `' password '` | :x:

### CODE REVIEW

File: dmutils/content_loader.py [hunk 1]
```diff
@@ -407,3 +410,44 @@ def read_yaml(yaml_file):
...
+
+        return s
+
+    def _strip_list(l):
+        return [_strip_values(v) for v in l]
+
+    def _strip_tuple(t):
+        return _strip_list(t)
+
+    def _strip_dict(d):
```
Comment (allait) [inline dmutils/content_loader.py]: Do we ever end up with dictionaries as values in `form_data`?

File: dmutils/content_loader.py [hunk 2]
```diff
@@ -175,6 +175,9 @@ def get_data(self, form_data):
         in the form data are ignored. Fields in the form data are parsed according
         to their type in the section data.
         """
+        # strip trailing and leading whitespace from form values
+        form_data = ImmutableMultiDict(_strip_values(form_data.to_dict(flat=False)))
```
Comment (allait) [inline dmutils/content_loader.py]: I think using `form_data.iteritems(multi=True)` here might simplify `_strip_values` a bit. My first though would be that:

    form_data = ImmutableMultiDict((k,v.strip()) for k,v in form_data.iteritems(multi=True))

should be enough, but I'd test this with assertions, features and benefits, pricing, and file uploads to check that it covers the edge cases.

**Comment by robyoung** [comment]
I couldn't think of any real reasons not to do this so I'm going to go with the majority. Code looks good :+1:

**Comment by pcraig3** [comment]
@TheDoubleK agrees with us as well.

:grinning:
